prpc/app.py

```diff
@@ -0,0 +1,85 @@
+# SPDX-License-Identifier: GPL-3.0-or-later
+import os
+import textwrap
+
+import flask
+import werkzeug.exceptions
+
+import prpc.db
+import prpc.errors
+
+metrics_bp = flask.Blueprint("metrics", __name__)
+
+
+def create_app():
+    """
+    Create a configured Flask application object.
+
+    :return: the Flask application object
+    :rtype: flask.Flask
+    """
+    app = flask.Flask(__name__)
+    # Use the in-memory SQLite database for tests
+    if os.environ.get("PRPC_TESTING", "").lower() == "true":
+        db_path = ":memory:?cache=shared"
+    else:
+        current_dir = os.path.abspath(os.path.curdir)
+        db_path = os.path.join(current_dir, "prpc.db")
+    app.config["PRPC_DB_PATH"] = os.environ.get("PRPC_DB_PATH", db_path)
+    app.config["PRPC_METRIC_PREFIX"] = os.environ.get("PRPC_METRIC_PREFIX", "unnamed")
+    app.register_blueprint(metrics_bp)
+
+    for code in werkzeug.exceptions.default_exceptions.keys():
+        app.register_error_handler(code, prpc.errors.json_error)
+    app.register_error_handler(prpc.errors.ValidationError, prpc.errors.json_error)
+
+    return app
+
+
+@metrics_bp.route("/metrics")
+def metrics():
+    """Display Prometheus metrics related to the Tekton Pipeline."""
+    prefix = flask.current_app.config["PRPC_METRIC_PREFIX"]
+    db_path = flask.current_app.config["PRPC_DB_PATH"]
+    with prpc.db.DBConnection(db_path) as db:
+        total_success = db.get_pipeline_run_count(success=True)
+        total_failed = db.get_pipeline_run_count(success=False)
+
+    metrics_content = textwrap.dedent(
+        f"""\
+        # HELP {prefix}_pipeline_run_total The total number of Tekton Pipeline runs.
+        # TYPE {prefix}_pipeline_run_total counter
+        {prefix}_pipeline_run_total{{status="success"}} = {total_success}
+        {prefix}_pipeline_run_total{{status="failed"}} {total_failed}
+        """
+    )
+    return flask.Response(metrics_content, content_type="text/plain; charset=utf-8")
+
+
+@metrics_bp.route("/metrics", methods=["POST"])
+def add_pipeline_run():
```

Comment:
I was thinking at the end of a Tekton pipeline run by using [finally](https://tekton.dev/vault/pipelines-master/pipelines/#adding-finally-to-the-pipeline). I have not tried it yet though.

Reply:
Hmm, I'm not sure if the pipeline run status is accessible in the "finally" block (thought it would be possible to set a variable in a last task).

I'm also not sure if "finally" works in case of pipeline/run timeout.

Reply:
I'll give it a shot and see what our options are.